nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -616,7 +675,187 @@ impl<'a> BlueprintBuilder<'a> {
+                    let zpool = &dataset.dataset.pool_name;
+                    let address = match zone.zone_type {
+                        BlueprintZoneType::Crucible(
+                            blueprint_zone_type::Crucible { address, .. },
+                        ) => Some(address),
+                        _ => None,
+                    };
+                    datasets_builder.ensure(
+                        DatasetName::new(zpool.clone(), dataset.kind),
+                        address,
+                        None,
+                        None,
+                        CompressionAlgorithm::Off,
+                    );
+                }
+            }
+
+            // TODO: Note that we also have datasets in "zone/" for propolis
+            // zones, but these are not currently being tracked by blueprints.
+
+            let expunges = datasets_builder.get_expungeable_datasets();
+            let removals = datasets_builder.get_removable_datasets();
+
+            let additions = datasets_builder
+                .new_datasets
+                .into_values()
+                .flat_map(|datasets| datasets.into_values().map(|d| (d.id, d)))
+                .collect::<BTreeMap<_, _>>();
+            let updates = datasets_builder
+                .updated_datasets
+                .into_values()
+                .flat_map(|datasets| {
+                    datasets.into_values().map(|dataset| (dataset.id, dataset))
+                })
+                .collect::<BTreeMap<DatasetUuid, _>>();
+            (additions, updates, expunges, removals)
+        };
+
+        if additions.is_empty()
+            && updates.is_empty()
+            && expunges.is_empty()
+            && removals.is_empty()
+        {
+            return Ok(EnsureMultiple::NotNeeded);
+        }
+        let added = additions.len();
+        let updated = updates.len();
+        // - When a dataset is expunged, for whatever reason, it is a part of
+        // "expunges". This leads to it getting removed from a sled.
+        // - When we know that we've safely destroyed all traces of the dataset,
+        // it becomes a part of "removals". This means we can remove it from the
+        // blueprint.
+        let expunged = expunges.len();
+        let removed = removals.len();
+
+        let datasets =
+            &mut self.datasets.change_sled_datasets(sled_id).datasets;
+
+        // Add all new datasets
+        datasets.append(&mut additions);
+
+        for config in datasets.values_mut() {
+            // Apply updates
+            if let Some(new_config) = updates.remove(&config.id) {
+                *config = new_config;
+            };
+
+            // Mark unused datasets as expunged.
+            //
+            // This indicates that the dataset should be removed from the database.
+            if expunges.contains(&config.id) {
+                config.disposition = BlueprintDatasetDisposition::Expunged;
+            }
+        }
```

Comment:
I think I need to start removing from `expunges` to do that, but I can make that change. Updated.

Reply:
I see the change to remove from `expunges`, but I don't see a check that they're both empty after this loop - could there be?

Reply:
Not sure how this didn't make it earlier, but:

https://github.com/oxidecomputer/omicron/blob/60ff0440b0d0d6e63ef5dd79dde67da27a0fbc37/nexus/reconfigurator/planning/src/blueprint_builder/builder.rs#L874-L878

Reply:
That's not the same thing, right? That breaks out of the loop if they are empty; I'm asking if we should add a check after the loop that _confirms_ they're empty (since the expectation is that they should be, I think?).

Reply:
Okay, I added that here: 7ee5cc5feeb4c0de392aace5cf56bd4f5dc177e1

I think this is dead code, though, because I don't know how we could make these values be non-empty

Reply:
I hope that is true of all of our "double check we made no mistakes or bail out" checks :) But sometimes we get surprised.